File: youzi_v2/frontend/src/api/customers.ts
import { api } from '@/api/client'

export type CustomerLang = 'zh' | 'en'

export interface Customer {
  id: string
  customerName: string
  customerLang: CustomerLang
  isVip: boolean
  note: string
  shipmentCount: number
  settlementMethod?: string | null
  settlementDay?: number | null
  createdTime: string
  updatedTime: string
}

export interface CustomerListResponse {
  items: Customer[]
  total: number
  limit: number
  offset: number
}

export interface CustomerSyncResult {
  added: number
  total: number
  fromShipments: number
}

export type CustomerPatch = {
  customerName?: string
  updateShipments?: boolean
  isVip?: boolean
  note?: string
  customerLang?: CustomerLang
  settlementMethod?: string | null
  settlementDay?: number | null
}

export type CustomerUpdateResult = Customer & {
  shipmentsUpdated?: number
}

export async function listCustomers(params?: {
  search?: string
  vipOnly?: boolean
  limit?: number
  offset?: number
}): Promise<CustomerListResponse> {
  const q: Record<string, string> = {}
  if (params?.search?.trim()) q.search = params.search.trim()
  if (params?.vipOnly === true) q.vipOnly = 'true'
  if (params?.vipOnly === false) q.vipOnly = 'false'
  if (params?.limit != null) q.limit = String(params.limit)
  if (params?.offset != null) q.offset = String(params.offset)
  return api<CustomerListResponse>('/api/v1/customers', { query: q })
}

export async function syncCustomersFromShipments(): Promise<CustomerSyncResult> {
  return api<CustomerSyncResult>('/api/v1/customers/sync-from-shipments', {
    method: 'POST',
  })
}

export async function createCustomer(
  customerName: string,
  options?: {
    note?: string
    isVip?: boolean
    customerLang?: CustomerLang
  },
): Promise<Customer> {
  return api<Customer>('/api/v1/customers', {
    method: 'POST',
    body: {
      customerName: customerName.trim(),
      note: options?.note?.trim() || '',
      isVip: options?.isVip ?? false,
      customerLang: options?.customerLang ?? 'zh',
    },
  })
}

export async function updateCustomer(
  id: string,
  patch: CustomerPatch,
): Promise<CustomerUpdateResult> {
  return api<CustomerUpdateResult>(`/api/v1/customers/${id}`, {
    method: 'PATCH',
    body: patch,
  })
}

export async function deleteCustomer(id: string): Promise<void> {
  await api(`/api/v1/customers/${id}`, { method: 'DELETE' })
}

File: youzi_v2/routers/customers.py
from fastapi import APIRouter

from ..context import *
from ..services.payment_reminder_rules import SETTLEMENT_MONTHLY

router = APIRouter()

@router.get("/api/v1/customers")
def list_customers(
    search: str | None = None,
    vip_only: bool | None = Query(None, alias="vipOnly"),
    limit: int = 200,
    offset: int = 0,
):
    return customers_repo.list_rows(
        search=search, vip_only=vip_only, limit=limit, offset=offset
    )

@router.post("/api/v1/customers/sync-from-shipments", response_model=CustomerSyncResult)
def sync_customers_from_shipments():
    """从运单表抓取全部去重客户名写入客户表（已存在的不覆盖 VIP 状态）。"""
    return CustomerSyncResult(**customers_repo.sync_from_shipments())

@router.post("/api/v1/customers", status_code=201)
def create_customer(body: CustomerIn):
    try:
        return customers_repo.create(
            body.customer_name,
            note=body.note,
            is_vip=body.is_vip,
            customer_lang=body.customer_lang,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

@router.patch("/api/v1/customers/{item_id}")
def update_customer(item_id: str, body: CustomerUpdateIn):
    try:
        if body.customer_name is not None:
            row = customers_repo.rename(
                item_id,
                body.customer_name,
                update_shipments=bool(body.update_shipments),
            )
        else:
            row = customers_repo.patch(
                item_id,
                is_vip=body.is_vip,
                note=body.note,
                customer_lang=body.customer_lang,
                settlement_method=body.settlement_method,
                settlement_day=body.settlement_day,
                clear_settlement_day=body.settlement_method is not None
                and body.settlement_method != SETTLEMENT_MONTHLY
                and body.settlement_day is None,
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if row is None:
        raise HTTPException(status_code=404, detail="客户不存在")
    return row

@router.delete("/api/v1/customers/{item_id}", status_code=204)
def delete_customer(item_id: str):
    if not customers_repo.delete(item_id):
        raise HTTPException(status_code=404, detail="客户不存在")
    return Response(status_code=204)
